File: JavaScript/Queue02.js
class Queue {
    constructor() {
        this.items = [];
    }

    // add element to the queue
    enqueue(element) {
        this.items.push(element);
        return element;
    }

    // remove element from the queue
    dequeue() {
        if(this.items.length > 0) {
            return this.items.shift();
        }
    }

    // view the last element
    peek() {
        return this.items[this.items.length - 1];
    }

    // check if the queue is empty
    isEmpty(){
    return this.items.length == 0;
    }

    length(){
        return this.items.length;
    }
}

const marvelMovies = new Queue()
console.log('Inserindo os elementos na Queue')
console.log('INPUT: '+marvelMovies.enqueue('Iron Man'))
console.log('INPUT: '+marvelMovies.enqueue('Captain America'))
console.log('INPUT: '+marvelMovies.enqueue('Incredible Hulk'))
console.log('LENGTH: '+marvelMovies.length())
console.log('Removendo os elementos da Queue')
console.log('NEXT: '+marvelMovies.peek())
console.log('OUTPUT: '+marvelMovies.dequeue())
console.log('LENGTH: '+marvelMovies.length())
console.log('NEXT: '+marvelMovies.peek())
console.log('OUTPUT: '+marvelMovies.dequeue())
console.log('LENGTH: '+marvelMovies.length())
console.log('NEXT: '+marvelMovies.peek())
console.log('OUTPUT: '+marvelMovies.dequeue())
console.log('LENGTH: '+marvelMovies.length())
console.log('NEXT: '+marvelMovies.peek())
console.log('Está vazio? '+marvelMovies.isEmpty())

File: Go/Queue02.go
package main

import (
	"fmt"
	"strconv"
)

type QueueTailHead struct {
	queue []string
}

func (c *QueueTailHead) Add(value string) string {
	c.queue = append(c.queue, value)
	return c.queue[len(c.queue)-1]
}

func (c *QueueTailHead) Remove() string {
	x := c.queue[0]
	c.queue = c.queue[1:]
	return x
}

func (c *QueueTailHead) Next() string {
	if len(c.queue) > 0 {
		return c.queue[c.Length()-1]
	} else {
		return "undefined"
	}
}

func (c *QueueTailHead) Empty() bool {
	return len(c.queue) == 0
}

func (c *QueueTailHead) Length() int {
	return len(c.queue)
}

func main() {
	var marvelMovies QueueTailHead
	fmt.Println("FIFO => Inserindo os elementos na Queue")
	fmt.Println("INPUT: " + marvelMovies.Add("Iron Man"))
	fmt.Println("INPUT: " + marvelMovies.Add("Captain America"))
	fmt.Println("INPUT: " + marvelMovies.Add("Incredible Hulk"))
	fmt.Println("LENGTH: " + strconv.Itoa(marvelMovies.Length()))
	fmt.Println("FIFO => Removendo os elementos da Queue")
	fmt.Println("OUTPUT: " + marvelMovies.Remove())
	fmt.Println("NEXT: " + marvelMovies.Next())
	fmt.Println("LENGTH: " + strconv.Itoa(marvelMovies.Length()))
	fmt.Println("OUTPUT: " + marvelMovies.Remove())
	fmt.Println("NEXT: " + marvelMovies.Next())
	fmt.Println("LENGTH: " + strconv.Itoa(marvelMovies.Length()))
	fmt.Println("OUTPUT: " + marvelMovies.Remove())
	fmt.Println("NEXT: " + marvelMovies.Next())
	fmt.Println("LENGTH: " + strconv.Itoa(marvelMovies.Length()))
	fmt.Println("Está vazio? " + strconv.FormatBool(marvelMovies.Empty()))
}
